### ISSUE TITLE
Profiling

### ISSUE THREAD

**Issue by Aidiakapi** [pkulchenko/ZeroBraneStudio#277]
When it comes to performance Lua is an odd bird. It's hybrid table implementation and all strings being interned makes it hard to sometimes see where the lost time is.

For performance testing it'd be great to have a profiler available. After all, first measure, then optimize.

I realize there are some Lua libraries for profiling out there, but they're mostly cumbersome to use, not flexible, and most essentially, not very accurate. I'm quite positive that in C(++), an implementation could be much better.

**Comment by pkulchenko** [comment]
Proof of concept: https://github.com/CapsAdmin/zbprofiler

**Comment by abigpotostew** [comment]
@pkulchenko Is that a heat diagram for percentage of overall time? I would also consider adding the number of times a line is run and related metrics.

**Comment by pkulchenko** [comment]
@abigpotostew, yes, but there is a limited number of gradations as it's done with markers and there are only few of them in the editor. Yes, number of times a line is run would be interesting to add as well.

**Comment by gschenck** [comment]
Take a look at
http://howl.io/blog/2016/05/26/introducing-aullar.html#aullar
as a replacement (?) for Scintillia.

**Comment by pkulchenko** [comment]
@gschenck, it's not related to this ticket, but I'll answer it briefly. I definitely follow Howl with interest and carefully reviewed their rationale for switching away from Scintilla. Here is why some of that rationale doesn't apply to this project:

1. Implementation of better indicators (flairs in Howl terminology) has been added to Scintilla 3.5.3 and included in the current version of the IDE under development.
2. Support for flexible lexing (including embedded lexers) will be available with the included support for LexLPEG (from Scintillua). It's also based on LPEG, so similar in may aspects to what Howl is doing.
3. Removing a dependency and keeping all development in one language (plus reduced executable size) is more important to Howl than to this project as we still need to use wxwidgets, which is a much larger library than the (already included in it) Scintilla library.
4. Transparency and variable height styles, while visually interesting, do not warrant the current component to be replaced.
5. This project is using some of the Scintilla functionality that is not used (and as the results is not available) in Howl, for example, invisible styles, which are used in markdown formatting in comments.

**Comment by pkulchenko** [comment]
Couple of notes on this: [ProFi profiler](https://github.com/mindreframer/ProFi.lua/blob/master/ProFi.lua), [sampling profiler for Lua 5.3](https://github.com/scott-intense/countprof), [interactive continuous Python profiler](https://github.com/what-studio/profiling), [profile format used by DProf](http://perldoc.perl.org/5.8.9/Devel/DProf.html#PROFILE-FORMAT), and couple of screenshots from Unreal profiler: [1](https://answers.unrealengine.com/storage/temp/75173-frontendfirstspawn.png) and [2](https://docs.unrealengine.com/latest/images/Engine/Performance/Profiler/MainEventGraphWindow.jpg).
